internal/api/core/middleware/proxy.go

```diff
@@ -0,0 +1,118 @@
+package middleware
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/perses/common/etcd"
+	"github.com/perses/perses/internal/api/interface/v1/datasource"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	proxyMatcher                       = regexp.MustCompile(`/proxy/datasources/([a-zA-Z-0-9_-]+)(/.*)`)
+	defaultTransport http.RoundTripper = &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			DualStack: true,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		TLSHandshakeTimeout: 10 * time.Second,
+		MaxIdleConns:        100,
+		IdleConnTimeout:     90 * time.Second,
+		ForceAttemptHTTP2:   true,
+	}
+)
+
+func Proxy(dao datasource.DAO) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			targetURL, err := buildProxy(c, dao)
+			if err != nil {
+				return err
+			}
+			if targetURL == nil {
+				return next(c)
+			}
+			return serveProxy(c, targetURL)
+		}
+	}
+}
+
+func buildProxy(c echo.Context, dao datasource.DAO) (*url.URL, error) {
+	req := c.Request()
+	requestPath := req.URL.Path
+	if !proxyMatcher.MatchString(requestPath) {
+		// request is not matching the proxy path. Probably it's a request to the API itself
+		// That's why no error is returned in order to let the request passed.
+		return nil, nil
+	}
+	logrus.Tracef("'%s' is a request to a datasource", requestPath)
+	matchingGroups := proxyMatcher.FindAllStringSubmatch(requestPath, -1)
+	if len(matchingGroups) > 1 || len(matchingGroups[0]) <= 1 {
+		return nil, echo.NewHTTPError(http.StatusBadGateway, "unable to forward the request to the datasource, request not properly formatted")
+	}
+	datasourceName := matchingGroups[0][1]
+	// getting the datasource object
+	dts, err := dao.Get(datasourceName)
+	if err != nil {
+		if etcd.IsKeyNotFound(err) {
+			logrus.Debugf("unable to find the Datasource '%s'", datasourceName)
+			return nil, echo.NewHTTPError(http.StatusBadGateway, fmt.Sprintf("unable to forward the request to the datasource '%s', datasource doesn't exist", datasourceName))
+		}
+		logrus.WithError(err).Errorf("unable to find the datasource '%s', something wrong with the database", datasourceName)
+		return nil, echo.NewHTTPError(http.StatusInternalServerError, "internal server error")
+	}
+	var path string
```

Comment:
nit: Maybe path should be `/` if there is nothing passed on the URL params?

Reply:
As per HTTP 1.1 RFC:
```
   Note that the absolute path
   cannot be empty; if none is present in the original URI, it MUST be
   given as "/" (the server root).
```

Relevant section:
https://datatracker.ietf.org/doc/html/rfc2616#section-5.1.2

Reply:
done